### ISSUE TITLE
Cell2Location on large datasets

### ISSUE THREAD

**Issue by LinearParadox** [BayraktarLab/cell2location#356]
Is there any way to manage memory useage on large datasets? For example, when you're approaching ~40000 spots and ~10000 genes, memory use becomes huge. Is there a way to train seperate conditions, and then somehow merge the results? I'm not sure if this would be a simple merge between anndata objects or if it's a more involved process.

**Comment by vitkl** [comment]
This is indeed a common problem. An approach we started using recently is splitting the data randomly (stratified by batch and possibly by anatomy annotation), training several models, then merging results.

**Comment by LinearParadox** [comment]
Is there any specific function you use to merge the data? Or is merging the anndata objects good enough? I'm not sure how I would merge several models, especially for some of the downstream analysis functions.

**Comment by vitkl** [comment]
You can concatenate anndata. There are two important parts to consider - cell abundance in `adata.obsm` and all other model parameters (+ essential record about input data) in `adata.uns['mod']`. Merging `adata.uns['mod']` correctly is going to be more complex but you can save them as different `adata.uns` slots.

As a technical note, merging `adata.uns['mod']` would require some way to properly combine location-independent parameters - which is straightforward for technical effects but can be less straightforward for prior factorisation of cell abundance.

**Comment by Rafael-Silva-Oliveira** [comment]
> This is indeed a common problem. An approach we started using recently is splitting the data randomly (stratified by batch and possibly by anatomy annotation), training several models, then merging results.

Is there any improvements coming soon in cell2location to take in the full data? That approach of splitting in batches removes any spatial information that might be used to do cell deconvolution

With the new VisiumHD being more proeminent, the number of spots or bins can range anywhere from 150k to 650k+, so when I test with cell2location, I always get memory errors.

And doing in batches also doesn't work:

![image](https://github.com/BayraktarLab/cell2location/assets/70116258/8043eacf-7214-40d7-a53a-516c2cc03649)

Any near future implementations to support datasets such as VisiumHD?

**Comment by Li-ZhiD** [comment]
Hi,
I am using cell2location to deconve large Stereo-seq data. Could you show

**Comment by vitkl** [comment]
@Li-ZhiD @Dillon214

Split the object into batches
```python
# for every "sample", sample with replacement from chunks to allocate 
# some locations from each batch to all training batches
chunk_size = 72_000
chunks = [i for i in range(int(np.ceil(adata_vis.n_obs / chunk_size)))]

adata_vis.obs['training_batch'] = 0
for sample in adata_vis.obs['sample'].unique():
    ind = adata_vis.obs['sample'].isin([sample])
    adata_vis.obs.loc[ind, 'training_batch'] = np.random.choice(
        chunks, size=ind.sum(), replace=True, p=None
    )
    
adata_vis_full = adata_vis.copy()
for k in ['means', 'stds', 'q05', 'q95']:
    adata_vis_full.obsm[f"{k}_cell_abundance_w_sf"] = np.zeros((adata_vis_full.n_obs, inf_aver.shape[1]))
    
adata_vis.obs['training_batch'].value_counts()
```

Train
```python
seed = 0
scvi.settings.seed = seed
np.random.seed(seed)

# submit this chunk as separate jobs
for batch in adata_vis.obs['training_batch'].unique():
    # create and train the model
    scvi_run_name = f'{run_name_global}_batch{batch}_seed{seed}'
    print(scvi_run_name)

    training_batch_index = adata_vis_full.obs['training_batch'].isin([batch])
    adata_vis = adata_vis_full[training_batch_index, :].copy()
    
    # prepare anndata for scVI model
    cell2location.models.Cell2location.setup_anndata(
        adata=adata_vis, batch_key="sample"
    )

    # train as normal
   
    # export posterior
    import pyro
    # In this section, we export the estimated cell abundance (summary of the posterior distribution).
    adata_vis = mod.export_posterior(
        adata_vis, sample_kwargs={
            'batch_size': int(np.ceil(adata_vis.n_obs / 4)), 'accelerator': 'gpu',
            "return_observed": False,
        },
        add_to_obsm=['q05', 'q50'],
        use_quantiles=True,
    )
```

Complete the full object with cell abundance estimates from batched analysis
```python
    # copy cell2location results to the main object
    for k in adata_vis_full.obsm.keys():
        adata_vis_full.obsm[k][training_batch_index, :] = adata_vis.obsm[k].copy()
    adata_vis_full.uns[f'mod_{batch}'] = adata_vis.uns['mod'].copy()
```

**Comment by floriankreten** [comment]
I have a cohort of 10x Visium slides, that give 20k spots and 10k genes in total (after filtering). Would you recommend to use the individual slides as batches, and simplify the above approach for this special use-case?

**Comment by vitkl** [comment]
>20k spots and 10k genes in total 

You should be able to analyse this on one A100 40GB GPU.

**Comment by Li-ZhiD** [comment]
Hi, 
My memory of GPU only has 12GB, and it may fall far short of the memory required for a 100K spot for a stereo-seq chip.
I tried CPU mode by setting
```
os.environ["THEANO_FLAGS"] = 'device=cpu,floatX=float32,openmp=True,force_device=True'
```
it doesn't work, still use GPU.

or tried setting "use_gpu=False"：
```
mod.train(max_epochs=30000,
          # train using full data (batch_size=None)
          batch_size=None,
          # use all data points in training because
          # we need to estimate cell abundance at all locations
          train_size=1,
          use_gpu=False
         )
```
an error raised:
```
Trainer.__init__() got an unexpected keyword argument 'use_gpu'
```

What can I try if I want to use cell2location (1.2.0)?  Thanks!

**Comment by vitkl** [comment]
Cell2location doesn't use theano and your code suggest that you are using the older version that did (2019). I would recommend using the latest version.

**Comment by Li-ZhiD** [comment]
Thank you!

**Comment by vitkl** [comment]
I would recommend to get access to GPUs with more memory (eg on various cloud platforms) because training on CPU is much slower and also more expensive because you need a lot of CPU resources for a very long time.

**Comment by EddieLv** [comment]
@vitkl Hi, when I train model with batch_size=None (full data input), it throw out error like:
OutOfMemoryError: CUDA out of memory. Tried to allocate 9.88 GiB. GPU 0 has a total capacity of 15.70 GiB of which 5.52 GiB is free. Including non-PyTorch memory, this process has 10.17 GiB memory in use. Of the allocated memory 9.89 GiB is allocated by PyTorch, and 8.49 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

but when I set batch_size=10000, it can run without error, **_I wonder if the change of batch_size could influence the final result?_**
mod.train(max_epochs=30000,
          # train using full data (batch_size=None)
          batch_size=10000,
          # use all data points in training because
          # we need to estimate cell abundance at all locations
          train_size=1
         )

**Comment by kimsjune** [comment]
Hi @vitkl,

On a related note to splitting the dataset and training, do you think there is a way to merge `adata_vis_full.uns[f'mod_{batch}']` into one unified `adata_vis_full.uns[mod]`? This is required for downstream NMF analysis.

I'm also curious if the dictionary keys under `.uns['mod']` have been updated. For example, I don't see `['spot_factors']` in my `adata_vis` unlike `X_data = adata_vis.uns['mod']['post_sample_q05']['spot_factors']` from [this tutorial](https://cell2location.readthedocs.io/en/latest/notebooks/cell2location_short_demo_downstream.html?highlight=run_colocation#3.-Automatic-discrete-region-identification-by-clustering-cell-densities).

Thank you.

**Comment by vitkl** [comment]
@EddieLv 
>but when I set batch_size=10000, it can run without error, I wonder if the change of batch_size could influence the final result?

Yes. You need to compare data splitting strategy https://github.com/BayraktarLab/cell2location/issues/356#issuecomment-2226993138 to using a different batch size. We don't recommend using batch sizes other than full data because small batch sizes substantially reduce accuracy and make it incredibly slow to train for 30k epochs. Since you are not maxing out largest GPUs on the market (80GB, 96GB) but only "small" 16GB, I would recommend using a different GPU model first - with eg 32GB or 40GB or 24GB.

**Comment by vitkl** [comment]
> I'm also curious if the dictionary keys under .uns['mod'] have been updated. For example, I don't see ['spot_factors'] in my adata_vis unlike X_data = adata_vis.uns['mod']['post_sample_q05']['spot_factors'] from [this tutorial](https://cell2location.readthedocs.io/en/latest/notebooks/cell2location_short_demo_downstream.html?highlight=run_colocation#3.-Automatic-discrete-region-identification-by-clustering-cell-densities).

Yes, the names changed several years ago. You can see the slots `adata_vis.uns['mod']['post_sample_q05'].keys()` and compare to model code/paper supplementary methods.

>On a related note to splitting the dataset and training, do you think there is a way to merge adata_vis_full.uns[f'mod_{batch}'] into one unified adata_vis_full.uns[mod]? This is required for downstream NMF analysis.

Latest version NMF code uses adata.obsm:
```python
run_colocation(
    ...
    train_args={
        'cell_abundance_obsm': 'q05_cell_abundance_w_sf'
    }
    ...
)
```

It is in principle possible to concatenate all obs-specific slots (like `adata_vis.uns['mod']['post_sample_q05']['w_sf']`) and compute average of global slots - but we have not seen as use case for this.

**Comment by kimsjune** [comment]
@vitkl 

Thank you for the super prompt response. 

[This line](https://github.com/BayraktarLab/cell2location/blob/8480e66ef2fd8a7112b18432424e033f540d3171/cell2location/run_colocation.py#L139) seems to expect `adata_vis.uns['mod']`. For now I just copied `adata_vis.uns['mod_0']` to it and `run_colocation()` seems to run. I suppose NMF is running based on a subset of all spots. For my dataset this represents about 1/3 of total (~ 160k spots).

**Comment by vitkl** [comment]
I don’t understand what you mean about the subset - the NMF code takes cell
abundance from `adata.obsm`. The line you copied takes cell types names.

**Comment by kimsjune** [comment]
Looking at the code a bit more, I see what you mean now. That line merely takes cell type names.

I've been following the code above for training in batches. I tried to use `batch_size = None` on a GH200 with 96GB RAM to no avail. 

As you wrote above,
> "...splitting the data randomly (stratified by batch and possibly by anatomy annotation), training several models, then merging results."

From this line, 
```python
    # copy cell2location results to the main object
    for k in adata_vis_full.obsm.keys():
        adata_vis_full.obsm[k][training_batch_index, :] = adata_vis.obsm[k].copy()
    adata_vis_full.uns[f'mod_{batch}'] = adata_vis.uns['mod'].copy()
```

My understanding is that the final `adata_vis_full` does not have `.uns['mod']`, but rather `.uns['mod_0'], .uns['mod_1']`, etc. But `run_colocation()` expects `.uns['mod']` to be in my `adata_vis_full`, so it's an easy fix. NMF is *not* running on any batch subset as I previously thought.

On a side note, I had to remove `.value()` from [the preceding line](https://github.com/BayraktarLab/cell2location/blob/8480e66ef2fd8a7112b18432424e033f540d3171/cell2location/run_colocation.py#L138) in my fork since `adata_vis.obsm["q05_cell_abundance_w_sf"]` is already an array. Could be a product of copying `adata_vis` to `adata_vis_full` in the for loop.

Thanks again for the feedback.

**Comment by whitneyt1** [comment]
> [@Li-ZhiD](https://github.com/Li-ZhiD) [@Dillon214](https://github.com/Dillon214)
> Complete the full object with cell abundance estimates from batched analysis
> 
>     # copy cell2location results to the main object
>     for k in adata_vis_full.obsm.keys():
>         adata_vis_full.obsm[k][training_batch_index, :] = adata_vis.obsm[k].copy()
>     adata_vis_full.uns[f'mod_{batch}'] = adata_vis.uns['mod'].copy()


I had to edit this code to address an error about subsetting series to this: 
# copy cell2location results to the main object
training_batch_index = adata_vis_full.obs['training_batch'].isin([1])

for k in adata_vis_full.obsm.keys():
    adata_vis_full[training_batch_index, :].obsm[k]=adata_vis.obsm[k].copy()

**Comment by xiaodeng6410** [comment]
Hey, we also encountered the same issue when using Spotiphy for cell deconvolution.
My strategy is to sort the spatial coordinates (x and y axes) and group spatially adjacent spots into the same batch. This allows us to split large batches efficiently. Compared to random sampling, which often results in a mosaic-like annotation, this spatially sorted splitting yields much better results.

```python
# Select marker genes
adata_sc_marker = adata_sc[:, marker_gene]
adata_st_marker = adata_st[:, marker_gene]

# Construct scRNA-seq reference
sc_ref = spotiphy.construct_sc_ref(adata_sc_marker, key_type=key_type)
spotiphy.sc_reference.plot_heatmap(adata_sc_marker, key_type, save=True, out_dir=results_folder)

print("Starting CUDA computation...")

# Sort spatial transcriptomics data by spatial coordinates
adata_st_marker = adata_st[:, marker_gene]
adata_st_marker2 = adata_st_marker.copy()
adata_st_marker2.obs = adata_st_marker2.obs.sort_values(by=['array_row', 'array_col'], ascending=[True, True])
adata_st_marker2 = adata_st_marker2[adata_st_marker2.obs.index, :]

# Step 1: Split the data into 10 parts
num_parts = 10
split_size = math.ceil(len(adata_st_marker2) / num_parts)
processed_parts = []

# Step 2: Process each split
for i in range(num_parts):
    start_idx = i * split_size
    end_idx = min((i + 1) * split_size, len(adata_st_marker2))
    
    # Extract current split
    test_part = adata_st_marker2[start_idx:end_idx, :]
    
    # Step 3: Run deconvolution on the current split
    device = 'cuda' if torch.cuda.is_available() else 'cpu'
    X = test_part.layers["counts"].toarray().astype(np.float32)
    
    cell_proportion = spotiphy.deconvolution.estimation_proportion(
        X, adata_sc_marker, sc_ref, type_list, key_type,
        n_epoch=8000, plot=True, batch_prior=1, device=device
    )
    
    # Annotate cell proportions
    test_part.obs[type_list] = cell_proportion
    processed_parts.append(test_part)
```

This is my code for splitting and processing large ST datasets with Spotiphy.

**Comment by xiaodeng6410** [comment]
> 有趣的建议，感谢分享。更好的结果是什么意思？这是在 VisiumHD 的背景下吗？
> 
> 在 Visium 样本之间随机拆分非常有用，因为您可以从每个样本中添加一些位置/点，以便模型可以学习如何集成数据集。

Thanks very much for clarifying! To provide some context, initially, we annotated and segmented the 8μm VisiumHD bins using Spotify. During that process, we encountered GPU memory limitations due to the large size of the dataset. Sorting the bins based on XY coordinates allowed us to circumvent this issue, and we observed that spatially ordered data indeed produced more accurate results for Spotify predictions.

However, after resolving these technical challenges, we realized Spotify didn't align well with our specific research interest in analyzing the Aβ niche. Consequently, we transitioned to the Cell2location workflow. While exploring Cell2location, we noted similar GPU memory limitations arising from large dataset sizes. Drawing parallels from our Spotify experience, we proposed a similar segmentation approach for Cell2location.

Your clarification highlights the important differences between these two methods: Spotify explicitly utilizes spatial relationships between neighboring spots, hence benefiting significantly from continuous spatial ordering. In contrast, Cell2location employs Bayesian inference integrating single-cell references and does not rely explicitly on spatial proximity, making random sampling more robust and appropriate for its analysis.

Thank you again for your detailed explanation and patience in addressing my earlier misunderstanding. Apologies for any confusion caused!

**Comment by cafferychen777** [comment]
Hi everyone,

Thank you for this valuable discussion on handling large spatial datasets. The batching approach shared by @vitkl is really helpful for working within memory constraints.

For those who continue to face scalability challenges with very large datasets (100k+ spots from Visium HD or Stereo-seq), we recently developed **FlashDeconv** as an additional option. It uses randomized linear algebra to achieve O(N) scaling, processing ~1M spots in about 3 minutes without GPU requirements.

We'd be grateful for any feedback from users working with large datasets:

- GitHub: https://github.com/cafferychen777/flashdeconv
- Paper: https://doi.org/10.64898/2025.12.22.696108
- Install: `pip install flashdeconv`

Of course, cell2location's Bayesian framework provides rigorous uncertainty quantification that's essential for many applications. Thanks to @vitkl and the team for the continued support and detailed workarounds!
